File: src/ios/PulsatePlugin.m
#import "PulsatePlugin.h"
#import <PULPulsate/PULPulsate.h>



@implementation PulsatePlugin

static PULPulsateManager *pulsateManager;
static NSDictionary *launchOptions;

- (void)pluginInitialize{
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(finishLaunching:) name:UIApplicationDidFinishLaunchingNotification object:nil];
}

- (void)finishLaunching:(NSNotification *)notification{
    launchOptions = notification.userInfo;
}

- (void)setAuthData:(CDVInvokedUrlCommand *)command{
    NSError *error;
    PULAuthorizationData *data = [[PULAuthorizationData alloc] initWithAppId:[command argumentAtIndex:0] andAppKey:[command argumentAtIndex:1]  validationError:&error];
    if(!error)
        pulsateManager = [PULPulsateFactory getInstanceWithAuthorizationData:data withLocationEnabled:YES withPushEnabled:YES withLaunchOptions:launchOptions withPulsateAppDelegate:YES andPulsateNotificationDelegate:YES error:&error];
    
    pulsateManager.unauthorizedDelegate = self;
    pulsateManager.badgeDelegate = self;
    
    launchOptions = nil;
}

- (void)startPulsateSession:(CDVInvokedUrlCommand *)command {
    self.sessionStartCallbackId = command.callbackId;
    [pulsateManager startPulsateSession:^(BOOL success, NSError* error) {
        if (success)
        {
            CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"sessionStarted"];
            [result setKeepCallbackAsBool:YES];
            [self.commandDelegate sendPluginResult:result callbackId:self.sessionStartCallbackId];
        }
        else
        {
            CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"sessionFailed"];
            [result setKeepCallbackAsBool:YES];
            [self.commandDelegate sendPluginResult:result callbackId:self.sessionStartCallbackId];
        }
    }];
}

- (void)startPulsateSessionForAlias:(CDVInvokedUrlCommand *)command {
    self.sessionStartCallbackId = command.callbackId;
    NSString *alias = [command argumentAtIndex:0];
    if(alias) {
        [pulsateManager startPulsateSessionForAlias:alias withListener:^(BOOL success, NSError* error) {
            if (success)
            {
                CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"sessionStarted"];
                [result setKeepCallbackAsBool:YES];
                [self.commandDelegate sendPluginResult:result callbackId:self.sessionStartCallbackId];
            }
            else
            {
                CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"sessionFailed"];
                [result setKeepCallbackAsBool:YES];
                [self.commandDelegate sendPluginResult:result callbackId:self.sessionStartCallbackId];
            }
        }];
    }
}

- (void)logout:(CDVInvokedUrlCommand *)command{
    self.logoutCallbackId = command.callbackId;
    [pulsateManager logout:^(BOOL success, NSError* error) {
        if (success)
        {
            CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"logoutSucess"];
            [result setKeepCallbackAsBool:YES];
            [self.commandDelegate sendPluginResult:result callbackId:self.logoutCallbackId];
        }
        else
        {
            CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"logoutFailed"];
            [result setKeepCallbackAsBool:YES];
            [self.commandDelegate sendPluginResult:result callbackId:self.logoutCallbackId];
        }
    }];
}

- (void)setNewThreadButtonEnabled:(CDVInvokedUrlCommand *)command{
    NSNumber *enabled = [command argumentAtIndex:0];
    if(enabled)
        [pulsateManager setNewThreadButtonEnabled:[enabled boolValue]];
}

- (void)sendLocationWithBeaconEvents:(CDVInvokedUrlCommand *)command{
    NSNumber *enabled = [command argumentAtIndex:0];
    if(enabled)
        [pulsateManager sendLocationWithBeaconEvents:[enabled boolValue]];
}

- (void)setLocationUpdatesEnabled:(CDVInvokedUrlCommand *)command{
    NSNumber *enabled = [command argumentAtIndex:0];
    if(enabled)
        [pulsateManager setLocationUpdatesEnabled:[enabled boolValue]];
}


- (void)setInAppNotificationEnabled:(CDVInvokedUrlCommand *)command{
    NSNumber *enabled = [command argumentAtIndex:0];
    if(enabled)
        [pulsateManager enableInAppNotification:[enabled boolValue]];
}

- (void)showLastInAppNotification:(CDVInvokedUrlCommand *)command{
    [pulsateManager showLastInAppNotification];
}

- (void)setPushNotificationEnabled:(CDVInvokedUrlCommand *)command{
    NSNumber *enabled = [command argumentAtIndex:0];
    if(enabled)
        [pulsateManager setPushNotificationEnabled:[enabled boolValue]];
}

- (void)setUserAuthorized:(CDVInvokedUrlCommand *)command{
    NSNumber *authorized = [command argumentAtIndex:0];
    if(authorized)
        [pulsateManager setUserAuthorized:[authorized boolValue]];
}

- (void)showLastUnauthorizedMessage:(CDVInvokedUrlCommand *)command{
    [pulsateManager showLastUnauthorizedMessage];
}

- (void)updateFirstName:(CDVInvokedUrlCommand *)command{
    NSString *firstName = [command argumentAtIndex:0];
    if(firstName)
        [pulsateManager updateFirstName:firstName];
}

- (void)updateLastName:(CDVInvokedUrlCommand *)command{
    NSString *lastName = [command argumentAtIndex:0];
    if(lastName)
        [pulsateManager updateLastName:lastName];
}

- (void)updateEmail:(CDVInvokedUrlCommand *)command{
    NSString *email = [command argumentAtIndex:0];
    if(email)
        [pulsateManager updateEmail:email];
}

- (void)updateGender:(CDVInvokedUrlCommand *)command{
    NSNumber *gender = [command argumentAtIndex:0];
    if(gender)
        [pulsateManager updateGender:[gender intValue]];
}

- (void)updateAge:(CDVInvokedUrlCommand *)command{
    NSNumber *age = [command argumentAtIndex:0];
    if(age)
        [pulsateManager updateAge:[age intValue]];
}

- (void)setPrivacy:(CDVInvokedUrlCommand *)command{
    NSNumber *privacy = [command argumentAtIndex:0];
    if(privacy)
        [pulsateManager setPrivacy:[privacy intValue]];
}

- (void)createAttributeWithString:(CDVInvokedUrlCommand *)command{
    NSString *attributeName = [command argumentAtIndex:0];
    NSString *attributeValue = [command argumentAtIndex:1];
    if(attributeName && attributeValue)
        [pulsateManager createAttribute:attributeName withString:attributeValue];
}

- (void)createAttributeWithFloat:(CDVInvokedUrlCommand *)command{
    NSString *attributeName = [command argumentAtIndex:0];
    NSNumber *attributeValue = [command argumentAtIndex:1];
    if(attributeName && attributeValue)
        [pulsateManager createAttribute:attributeName withFloat:[attributeValue floatValue]];
}

- (void)createAttributeWithInt:(CDVInvokedUrlCommand *)command{
    NSString *attributeName = [command argumentAtIndex:0];
    NSNumber *attributeValue = [command argumentAtIndex:1];
    if(attributeName && attributeValue)
        [pulsateManager createAttribute:attributeName withInteger:[attributeValue intValue]];
}

- (void)createAttributeWithBool:(CDVInvokedUrlCommand *)command{
    NSString *attributeName = [command argumentAtIndex:0];
    NSNumber *attributeValue = [command argumentAtIndex:1];
    if(attributeName && attributeValue)
        [pulsateManager createAttribute:attributeName withBoolean:[attributeValue boolValue]];
}

- (void)createAttributeWithDate:(CDVInvokedUrlCommand *)command{
    NSString *attributeName = [command argumentAtIndex:0];
    NSNumber *attributeValue = [command argumentAtIndex:1];
    
    if(attributeName && attributeValue){
        NSDate *formattedDate = [NSDate dateWithTimeIntervalSince1970:[attributeValue doubleValue] / 1000];
        [pulsateManager createAttribute:attributeName withDate:formattedDate];
    }
}

- (void)incrementCounter:(CDVInvokedUrlCommand *)command{
    NSString *counterName = [command argumentAtIndex:0];
    NSNumber *value = [command argumentAtIndex:1];
    
    if (strcmp([value objCType], @encode(float)) == 0 || strcmp([value objCType], @encode(double)) == 0) {
        [pulsateManager incrementFloatAttribute:counterName withFloat:[value floatValue]];
    } else if (strcmp([value objCType], @encode(int)) == 0 || strcmp([value objCType], @encode(long)) == 0) {
        [pulsateManager incrementIntegerAttribute:counterName withInteger:[value intValue]];
    }
}

- (void)decrementCounter:(CDVInvokedUrlCommand *)command{
    NSString *counterName = [command argumentAtIndex:0];
    NSNumber *value = [command argumentAtIndex:1];
    
    if (strcmp([value objCType], @encode(float)) == 0 || strcmp([value objCType], @encode(double)) == 0) {
        [pulsateManager decrementFloatAttribute:counterName withFloat:[value floatValue]];
    } else if (strcmp([value objCType], @encode(int)) == 0 || strcmp([value objCType], @encode(long)) == 0) {
        [pulsateManager decrementIntegerAttribute:counterName withInteger:[value intValue]];
    }
}

- (void)createEvent:(CDVInvokedUrlCommand *)command{
    NSString *event = [command argumentAtIndex:0];
    if(event)
        [pulsateManager createEvent:event];
}

- (void)forceAttributeSync:(CDVInvokedUrlCommand *)command{
    [pulsateManager forceAttributeSync];
}

- (void)showFeed:(CDVInvokedUrlCommand *)command{
    UINavigationController* pulsateFeedNavController = [pulsateManager getFeedNavigationController];
    [self.viewController presentViewController:pulsateFeedNavController animated:YES completion:nil];
}

- (void)useInitialsForUserAvatar:(CDVInvokedUrlCommand *)command{
    NSNumber *useInitials = [command argumentAtIndex:0];
    if(useInitials)
        [pulsateManager useInitialsForUserAvatar:[useInitials boolValue]];
}

- (void)setSmallInAppNotificationDuration:(CDVInvokedUrlCommand *)command{
    NSNumber *duration = [command argumentAtIndex:0];
    if(duration)
        [pulsateManager useInitialsForUserAvatar:[duration intValue]];
}

- (void)setAuthDataIOS:(CDVInvokedUrlCommand *)command{
    NSError *error;
    PULAuthorizationData *data = [[PULAuthorizationData alloc] initWithAppId:[command argumentAtIndex:0] andAppKey:[command argumentAtIndex:1]  validationError:&error];
    
    NSNumber *location = [command argumentAtIndex:2];
    NSNumber *push = [command argumentAtIndex:3];
    NSNumber *app_delegate = [command argumentAtIndex:4];
    NSNumber *notification_delegate = [command argumentAtIndex:5];
    
    if(!error)
        pulsateManager = [PULPulsateFactory getInstanceWithAuthorizationData:data withLocationEnabled:[location boolValue] withPushEnabled:[push boolValue] withLaunchOptions:launchOptions withPulsateAppDelegate:[app_delegate boolValue] andPulsateNotificationDelegate:[notification_delegate boolValue] error:&error];
    
    pulsateManager.unauthorizedDelegate = self;
    pulsateManager.badgeDelegate = self;
    
    launchOptions = nil;
}

- (void)getDeviceGuidIOS:(CDVInvokedUrlCommand *)command{
    self.guidCallbackId = command.callbackId;
    NSString *guid = [pulsateManager getDeviceGuid];
    
    CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:guid];
    [result setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:result callbackId:self.guidCallbackId];
}

- (void)getBadgeCountIOS:(CDVInvokedUrlCommand *)command{
    [pulsateManager getBadgeCount];
}

- (void)onUnauthorizedActionIOS:(CDVInvokedUrlCommand *)command{
    self.unauthorizedCallbackId = command.callbackId;
}

- (void)onBadgeUpdatedIOS:(CDVInvokedUrlCommand *)command{
    self.badgeUpdatedCallbackId = command.callbackId;
}

- (void)onBadgeDecrementByIOS:(CDVInvokedUrlCommand *)command{
    self.badgeDecrementByCallbackId = command.callbackId;
}

- (void)onBadgeIncrementByIOS:(CDVInvokedUrlCommand *)command{
    self.badgeIncrementByCallbackId = command.callbackId;
}

- (void)unauthorizedAction:(NSString *)action {
    if(self.unauthorizedCallbackId){
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:action];
        [result setKeepCallbackAsBool:YES];
        [self.commandDelegate sendPluginResult:result callbackId:self.unauthorizedCallbackId];
    }
}

- (void)badgeUpdated:(NSInteger)badgeNumber {
    if(self.badgeUpdatedCallbackId){
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsNSInteger:badgeNumber];
        [result setKeepCallbackAsBool:YES];
        [self.commandDelegate sendPluginResult:result callbackId:self.badgeUpdatedCallbackId];
    }
}

- (void)badgeDecrementBy:(NSInteger)badgeDecrement totalCount:(NSInteger)badgeNumber {
    if(self.badgeDecrementByCallbackId){
        NSDictionary *response = @{@"totalCount" : @(badgeNumber), @"badgeDecrement": @(badgeDecrement)};
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:response];
        [result setKeepCallbackAsBool:YES];
        [self.commandDelegate sendPluginResult:result callbackId:self.badgeDecrementByCallbackId];
    }
}

- (void)badgeIncrementBy:(NSInteger)badgeIncrement totalCount:(NSInteger)badgeNumber {
    if(self.badgeIncrementByCallbackId){
        NSDictionary *response = @{@"totalCount" : @(badgeNumber), @"badgeIncrement": @(badgeIncrement)};
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:response];
        [result setKeepCallbackAsBool:YES];
        [self.commandDelegate sendPluginResult:result callbackId:self.badgeIncrementByCallbackId];
    }
}
@end

File: src/android/PulsatePlugin.java
package com.pulsatehq.cordova;


import com.pulsatehq.external.pulsate.listener.IPulsateRequestListener;
import com.pulsatehq.external.pulsate.listener.IPulsateUnreadCountUpdateListener;
import com.pulsatehq.external.pulsate.manager.IPulsateManager;
import com.pulsatehq.internal.util.AuthorizationData;

import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import java.util.ArrayList;
import java.util.Date;


public class PulsatePlugin extends CordovaPlugin {
    
    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
    }
    
    @Override
    public boolean execute(String action, JSONArray args, final CallbackContext callbackContext) throws JSONException {
        IPulsateManager manager = ((MyApplication) this.cordova.getActivity().getApplication()).getPulsate();
        switch (action) {
            case "setAuthorizationData":
                AuthorizationData data = new AuthorizationData(args.getString(0), args.getString(1), args.getString(2));
                manager.setAuthorizationData(data);
                callbackContext.success();
                return true;
            case "startPulsateSession":
                manager.startPulsateSession(new IPulsateRequestListener() {
                    @Override
                    public void onSucess() {
                        callbackContext.success();
                    }

                    @Override
                    public void onError(Throwable e) {
                        callbackContext.error(e.getMessage());
                    }
                });
                return true;
            case "startPulsateSessionForAlias":
                manager.startPulsateSessionForAlias(args.getString(0), new IPulsateRequestListener() {
                    @Override
                    public void onSucess() {
                        callbackContext.success();
                    }

                    @Override
                    public void onError(Throwable e) {
                        callbackContext.error(e.getMessage());
                    }
                });
                return true;
            case "logout":
                manager.logoutCurrentAlias(new IPulsateRequestListener() {
                    @Override
                    public void onSucess() {
                        callbackContext.success();
                    }

                    @Override
                    public void onError(Throwable e) {
                        callbackContext.error(e.getMessage());
                    }
                });
                return true;
            case "setLocationUpdatesEnabled":
                manager.setLocationUpdatesEnabled(args.getBoolean(0));
                callbackContext.success();
                return true;
            case "setInAppNotificationEnabled":
                manager.setInAppNotificationEnabled(args.getBoolean(0));
                callbackContext.success();
                return true;
            case "showLastInAppNotification":
                manager.showLastInAppNotification();
                callbackContext.success();
                return true;
            case "setPushNotificationEnabled":
                manager.setPushNotificationEnabled(args.getBoolean(0));
                callbackContext.success();
                return true;
            case "setUserAuthorized":
                manager.setUserAuthorized(args.getBoolean(0));
                callbackContext.success();
                return true;
            case "showLastUnauthorizedMessage":
                manager.showLastUnauthorizedMessage();
                callbackContext.success();
                return true;
            case "setNewThreadButtonEnabled":
                manager.setNewThreadButtonEnabled(args.getBoolean(0));
                callbackContext.success();
                return true;
            case "sendLocationWithBeaconEvents":
                manager.sendLocationWithBeaconEvents(args.getBoolean(0));
                callbackContext.success();
                return true;
            case "updateFirstName":
                String firstName = args.getString(0);
                manager.updateFirstName(firstName);
                callbackContext.success();
                return true;
            case "updateLastName":
                String lastName = args.getString(0);
                manager.updateLastName(lastName);
                callbackContext.success();
                return true;
            case "updateEmail":
                String email = args.getString(0);
                manager.updateEmail(email);
                callbackContext.success();
                return true;
            case "updateAge":
                Integer age = args.getInt(0);
                manager.updateAge(age.toString());
                callbackContext.success();
                return true;
            case "updateGender":
                Integer gender = args.getInt(0);
                if (gender == 1)
                    gender = 0;
                else
                    gender = 1;
                manager.updateGender(gender);
                callbackContext.success();
                return true;
            case "setPrivacy":
                Integer privacy = args.getInt(0);
                if (privacy == 1)
                    privacy = 0;
                else
                    privacy = 1;
                manager.setPrivacy(privacy);
                callbackContext.success();
                return true;
            case "createAttributeString": {
                String attributeName = args.getString(0);
                String attributeValue = args.getString(1);
                manager.createAttribute(attributeName, attributeValue);
                callbackContext.success();
                return true;
            }
            case "createAttributeInt": {
                String attributeName = args.getString(0);
                Integer attributeValue = args.getInt(1);
                manager.createAttribute(attributeName, attributeValue);
                callbackContext.success();
                return true;
            }
            case "createAttributeFloat": {
                String attributeName = args.getString(0);
                Double attributeValue = args.getDouble(1);
                manager.createAttribute(attributeName, attributeValue.floatValue());
                callbackContext.success();
                return true;
            }
            case "createAttributeBool": {
                String attributeName = args.getString(0);
                Boolean attributeValue = args.getBoolean(1);
                manager.createAttribute(attributeName, attributeValue);
                callbackContext.success();
                return true;
            }
            case "createAttributeDate": {
                String attributeName = args.getString(0);
                long attributeValue = args.getLong(1);
                Date date = new Date(attributeValue);
                manager.createAttribute(attributeName, date);
                callbackContext.success();
                return true;
            }
            case "createEvent":
                manager.createEvent(args.getString(0));
                callbackContext.success();
                return true;
            case "createEvents":
                ArrayList<String> listdata = new ArrayList<String>();
                for (int i = 0; i < args.length(); i++)
                    listdata.add(args.getString(i));
                manager.createEvents(listdata);
                callbackContext.success();
                return true;
            case "incrementCounter": {
                String counterName = args.getString(0);
                Object value = args.get(1);
                if (value instanceof Integer)
                    manager.incrementCounter(counterName, (Integer) value);
                callbackContext.success();
                return true;
            }
            case "decrementCounter": {
                String counterName = args.getString(0);
                Object value = args.get(1);
                if (value instanceof Integer)
                    manager.decrementCounter(counterName, (Integer) value);
                callbackContext.success();
                return true;
            }
            case "forceAttributeSync":
                manager.forceAttributeSync();
                callbackContext.success();
                return true;
            case "showFeed":
                manager.showFeed();
                callbackContext.success();
                return true;
            case "setUnreadCountUpdateListener":
                manager.setUnreadCountUpdateListener(new IPulsateUnreadCountUpdateListener() {
                    @Override
                    public void onUnreadCountUpdate(int unread) {
                        PluginResult result = new PluginResult(PluginResult.Status.OK, unread);
                        result.setKeepCallback(true);
                        callbackContext.sendPluginResult(result);
                    }
                });
                return true;
        }
        return false;
    }
}
